Add a `MAGIC` associated const for BinRead types with a magic attribute

MrNbaYoh opened the pull request:
Sometimes you want to get the magic value of your type.
By automatically defining an associated const we avoid the need for external definitions and/or redundancy.
We should probably move the code that creates the `magic_definition` somewhere else though.

jam1garner commented:
@MrNbaYoh I feel like `MAGIC` could have a non-zero rate for name collision. Could we instead throw it behind a trait (with the added benefit it could then be abstracted over). Outside of that I definitely think this is a good addition.

Possible trait definition:

```rust
trait HasMagic {
    type MagicType;
    const MAGIC: Self::MagicType;
}
```

That way users can opt-in to it by importing the trait (make sure any macro expansion doesn't import the trait, just uses the fully-qualified path (see the [sanitization](https://github.com/jam1garner/binrw/blob/master/binrw_derive/src/codegen/sanitization.rs) module if you aren't familiar)).

Also if you wouldn't mind adding a test (just throw a bunch of generics and lifetimes onto a struct with magic then make sure you can assert_eq the contents of `<T as HashMagic>::MAGIC`) that'd be appreciated--although if you don't feel like writing one I can handle that

MrNbaYoh commented:
@jam1garner I've changed the implementation, added the `HasMagic` trait and moved the code generation to its own file. I wrote a basic test with no generics/lifetimes (not sure how I'd add lifetimes to a `BinRead` type). I haven't written the doc for the trait yet.
